Raster Planning improvments

marrts opened the pull request:
Added several features with the aim of improving motion planning

- Ability to configurably add buffer to ends of raster strips (to avoid getting to close to the edge of the part)

- Ability to define a "sphere of reachability" from the base of the robot to prevent the robot from over extending

- Ability to define minimum absolute joint values for each joint when going through descartes (helps prevent singularities)

- Added new parameter to split toolpaths based on rotation rate between adjacent points (results on better individual rasters and gets rid of points on areas of sharp curvature)

- Added ability for OMPL to retry with smaller constraints if start or end state is thought to be in collision (prevents rare error that would cause OMPL to fail after trajopt smoothed a surface path)

- Added various helper functions to path_processing_tools

- Removed unnecessary media change_pose parameter that was no longer used

jrgnicho commented on crs_application/config/main/motion_planning/MP_config.yaml:
```diff
@@ -11,16 +11,19 @@ general:
   global_descartes: true
 descartes:
-  axial_step: 0.4
+  axial_step: 0.1
```
Does this increase planning time significantly?

marrts replied:
It did increase it, but not a major difference on my machine

jrgnicho commented on crs_application/config/main/motion_planning/MP_config.yaml:
```diff
@@ -11,16 +11,19 @@ general:
+  axial_step: 0.1
   collision_safety_margin: 0.00
   additional_tool_offset: [0.0, 0.0, 0.004, 0.0, 0.0, 0.0]
+  joint_min_vals: [-0.1, -0.1, -0.4, -0.1, 0.1, -0.1]
```
How are these used?

marrts replied:
These are used to check for valid joint states in descartes and are minimum allowable absolute values for each joint, so anything negative means that the corresponding joint is not restricted at all. The only thing this particular parameter does right now is ensure that wrist_2_joint has a value outside of -0.1 to 0.1 to avoid wrist singularities.

jrgnicho commented on crs_application/include/crs_application/common/config.h:
```diff
@@ -105,6 +104,7 @@ struct PartRegistrationConfig
   std::string toolpath_file;
   std::array<double, 6> seed_pose;
   std::array<double, 6> simulation_pose;
+  double waypoint_edge_buffer;
```
Shouldn't this be a parameter specific to motion planning?

marrts replied:
It's here because this is where the toolpath is loaded and it is immediately applied to the loaded toolpath

jrgnicho commented on crs_motion_planning/include/crs_motion_planning/path_planning_utils.h:
```diff
@@ -47,6 +47,8 @@ struct descartesConfig
+  std::vector<double> joint_min_vals;
 };
 
 struct trajoptSurfaceConfig
```
Just realized that this struct does not adhere to naming conventions, first character should be in capitalized

jrgnicho reviewed (commented):
I'm good with this PR for the most part, I'll build and run just to make sure it ok to merge

jrgnicho commented:
Looks line clang formatting fail however

jrgnicho commented:
@marrts it fails to load the part because none of the current crs.yaml files conform to the new format

marrts commented:
> @marrts it fails to load the part because none of the current crs.yaml files conform to the new format

The crs.yaml files in the repo should be updated, I did not update the ones in crs_data though.

jrgnicho commented:
> > @marrts it fails to load the part because none of the current crs.yaml files conform to the new format
> 
> The crs.yaml files in the repo should be updated, I did not update the ones in crs_data though.

I see, the ones in the crs_process_data are the ones that are loaded by the application first, it only uses the ones in crs_application if it doesn't find it in crs_process_data

marrts commented:
I have updated the part_5 crs.yaml to match the one I was using.

I removed the change_pose parameter and added the toolpath_edge_buffer one. Would you like me to go through and do this with all of the crs.yaml files?

jrgnicho commented:
I think I can make a quick fix to the parsing to have backwards compatibility, let me make the change and I'll push to your branch

jrgnicho commented:
@marrts I just made a PR to your branch
https://github.com/marrts/collaborative-robotic-sanding/pull/3

The motion planning node seems to fail and terminate for some reason on part_2,it works fine on part 5 however

jrgnicho commented:
It seems the motion planning crash is unique to this branch. I ran gdb and this is the output from the terminal
```
Info:    ParallelPlan::solve(): Solution found by one or more threads in 0.482616 seconds
[Thread 0x7fff0b7f6700 (LWP 24539) exited]
Info:    SimpleSetup: Path simplification took 0.240906 seconds and changed from 17 to 49 states
[INFO] [crs.motion_planning_server]: OMPL SEED GENERATED
[INFO] [crs.motion_planning_server]: SETTING UP TRAJOPT CONFIG
[INFO] [crs.motion_planning_server]: OPTIMIZING WITH TRAJOPT
[Thread 0x7fff0ffff700 (LWP 24530) exited]
[Thread 0x7fff24ff9700 (LWP 24529) exited]
[Thread 0x7fff26ffd700 (LWP 24528) exited]

.
.
.
Program terminated with signal SIGKILL, Killed.
The program no longer exists.
bpmpd_caller: /home/jnicho/ros2/eloquent/crs_ws/src/trajopt_ros/trajopt_sco/include/trajopt_sco/bpmpd_io.hpp:36: void bpmpd_io::ser(int, T&, bpmpd_io::SerMode) [with T = char]: Assertion `n == sizeof(T)' failed.
```

marrts commented:
I'm not exactly sure what to make of that error, it hasn't happened to me before and I just also tested with part_2

jrgnicho commented:
OK, well if you can't reproduce it might be just on my end.  I'll merge.  Thanks
